Maple2.File.Parser/AiParser.cs

```diff
@@ -10,24 +10,265 @@
+                case "Int32[]":
+                    int[] intValues = Array.ConvertAll(attrib.Value.Split(",", StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries), int.Parse);
+
+                    field.SetValue(outObject, intValues);
+                    break;
+                default:
+                    if (field.FieldType.IsArray && field.FieldType.GetElementType().IsEnum) {
+                        Type? nestedType = field.FieldType.GetElementType();
+                        if (nestedType is null) {
+                            throw new InvalidDataException($"unhandled type of {field.FieldType.Name} on value '{attrib.Value}' for {typeof(T).Name} member {field.Name}");
+                        }
+                        // generates an object[] array of the enums. i couldn't find a way to generate the actual underlying enum type array directly because it kept resolving to object[]
+                        var enumObjectValues = Array.ConvertAll(attrib.Value.Split(",", StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries), value => System.Enum.Parse(nestedType, value));
+                        // forcibly create an enum array and copy values over so that field.SetValue is happy & copies the array over to the member
+                        dynamic enumValues = Array.CreateInstance(field.FieldType.GetElementType(), enumObjectValues.GetLength(0));
+                        for (int i = 0; i < enumObjectValues.Length; ++i) {
+                            enumValues[i] = (dynamic)enumObjectValues[i];
+                        }
+                        field.SetValue(outObject, enumValues);
+                    }
+                    else if (field.FieldType.IsEnum) {
+                        field.SetValue(outObject, System.Enum.Parse(field.FieldType, attrib.Value));
+                    }
+                    else {
+                        throw new InvalidDataException($"unhandled type of {field.FieldType.Name} on value '{attrib.Value}' for {typeof(T).Name} member {field.Name}");
+                    }
+                    break;
+            }
+        }
     }
 
     public IEnumerable<(string aiName, NpcAi Data)> Parse() {
         foreach (PackFileEntry entry in xmlReader.Files.Where(entry => entry.Name.StartsWith("AI/"))) {
             string sanitized = Sanitizer.SanitizeAi(xmlReader.GetString(entry));
-            var root = aiSerializer.Deserialize(XmlReader.Create(new StringReader(sanitized))) as NpcAi;
 
-            Debug.Assert(root != null);
+            XmlDocument document = new XmlDocument();
+            var nodeStack = new List<(int StackIndex, XmlNodeList Nodes, object OutNode)>();
+            NpcAi root = new NpcAi();
+
+            document.LoadXml(sanitized);
+
+            nodeStack.Add((0, document.ChildNodes, root));
+
+            while (nodeStack.Count > 0) {
+                (int StackIndex, XmlNodeList Nodes, object OutNode) = nodeStack.Last();
```

Comment:
local variables should be lowercase, this includes in tuples. Exception is when it's not the actual variable name, for example:

```cs
(int Field1, int Field2) obj = ...;

obj.Field1 // this is okay
```